src/mumble/GlobalShortcut_win.cpp

```diff
@@ -130,576 +52,356 @@ GlobalShortcutWin::~GlobalShortcutWin() {
-	//
-	// In the future, if there are more devices in the
-	// blacklist, we need a more structured aproach.
-	{
-		if (id->vendor_id == 0x262A) {
-			qWarning("GlobalShortcutWin: rejected blacklisted device %s (GUID: %s, PGUID: %s, VID: 0x%.4x, PID: "
-					 "0x%.4x, TYPE: 0x%.8lx)",
-					 qPrintable(id->name), qPrintable(id->vguid.toString()), qPrintable(id->vguidproduct.toString()),
-					 id->vendor_id, id->product_id, static_cast< unsigned long >(pdidi->dwDevType));
-			delete id;
-			return DIENUM_CONTINUE;
-		}
+	if (GetRawInputDeviceInfo(deviceHandle, RIDI_PREPARSEDDATA, nullptr, &size) != 0) {
+		return m_devices.end();
 	}
 
-	foreach (InputDevice *dev, cbgsw->qhInputDevices) {
-		if (dev->guid == id->guid) {
-			delete id;
-			return DIENUM_CONTINUE;
-		}
+	device.data.resize(size);
+	auto data = reinterpret_cast< PHIDP_PREPARSED_DATA >(&device.data[0]);
+	if (GetRawInputDeviceInfo(deviceHandle, RIDI_PREPARSEDDATA, data, &size) <= 0) {
+		return m_devices.end();
 	}
 
-	if (HAS_FAILED(hr = cbgsw->pDI->CreateDevice(pdidi->guidInstance, &id->pDID, nullptr)))
-		qFatal("GlobalShortcutWin: CreateDevice: %lx", hr);
-
-	if (HAS_FAILED(hr = id->pDID->EnumObjects(EnumDeviceObjectsCallback, static_cast< void * >(id), DIDFT_BUTTON)))
-		qFatal("GlobalShortcutWin: EnumObjects: %lx", hr);
-
-	if (id->qhNames.count() > 0) {
-		QList< DWORD > types = id->qhNames.keys();
-		std::sort(types.begin(), types.end());
-
-		int nbuttons = types.count();
-		STACKVAR(DIOBJECTDATAFORMAT, rgodf, nbuttons);
-		DIDATAFORMAT df;
-		ZeroMemory(&df, sizeof(df));
-		df.dwSize     = sizeof(df);
-		df.dwObjSize  = sizeof(DIOBJECTDATAFORMAT);
-		df.dwFlags    = DIDF_ABSAXIS;
-		df.dwDataSize = (nbuttons + 3) & (~0x3);
-		df.dwNumObjs  = nbuttons;
-		df.rgodf      = rgodf;
-		for (int i = 0; i < nbuttons; i++) {
-			ZeroMemory(&rgodf[i], sizeof(DIOBJECTDATAFORMAT));
-			DWORD dwType            = types[i];
-			DWORD dwOfs             = i;
-			rgodf[i].dwOfs          = dwOfs;
-			rgodf[i].dwType         = dwType;
-			id->qhOfsToType[dwOfs]  = dwType;
-			id->qhTypeToOfs[dwType] = dwOfs;
-		}
+	HIDP_CAPS caps;
+	if (HidP_GetCaps(data, &caps) != HIDP_STATUS_SUCCESS) {
+		return m_devices.end();
+	}
 
-		if (HAS_FAILED(hr = id->pDID->SetCooperativeLevel(mumble_mw_hwnd, DISCL_NONEXCLUSIVE | DISCL_BACKGROUND)))
-			qFatal("GlobalShortcutWin: SetCooperativeLevel: %lx", hr);
+	USHORT nCaps = caps.NumberInputButtonCaps;
+	std::vector< HIDP_BUTTON_CAPS > buttonCaps(nCaps);
+	if (HidP_GetSpecificButtonCaps(HidP_Input, HID_USAGE_PAGE_BUTTON, 0, 0, &buttonCaps[0], &nCaps, data)
+		!= HIDP_STATUS_SUCCESS) {
+		return m_devices.end();
+	}
 
-		if (HAS_FAILED(hr = id->pDID->SetDataFormat(&df)))
-			qFatal("GlobalShortcutWin: SetDataFormat: %lx", hr);
+	device.usageRange.first  = buttonCaps[0].Range.UsageMin;
+	device.usageRange.second = buttonCaps[0].Range.UsageMax;
+	device.buttons.resize(device.usageRange.second - device.usageRange.first + 1);
```

Comment:
Why `+1`?

Reply:
From https://www.codeproject.com/articles/185522/using-the-raw-input-api-to-process-joystick-input:

> `HidP_GetButtonCaps` returns the capabilities for every type of [HID control](http://msdn.microsoft.com/en-us/library/ff539879(VS.85).aspx) of our joystick, that is, its Usage. `pButtonCaps[0].UsagePage` indicates the usage of the first set of HID controls (see Table 1 of [HID Usage](http://www.usb.org/developers/devclass_docs/Hut1_12.pdf) Tables). `UsageMin` and `UsageMax` indicate the lower and upper bound of usage range, i.e. the range of indicies that is returned by the `HIDClass` driver for the buttons of the joystick. The number of buttons is then the difference of `UsageMin` and `UsageMax` plus one.